src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                        else if (stream == MinishellStream.Debug ||
+                                 stream == MinishellStream.Verbose ||
+                                 stream == MinishellStream.Warning)
+                        {
+                            //Convert to string
+                            try
+                            {
+                                obj = LanguagePrimitives.ConvertTo(obj, typeof(string), CultureInfo.InvariantCulture);
+                            }
+                            catch (PSInvalidCastException)
                             {
-                                startPosition.Y = 0;
+                                continue;
                             }
+                        }
+                        result.Add(new ProcessOutputObject(obj, stream));
+                    }
+                }
+            }
+            catch (XmlException originalException)
+            {
+                string template = NativeCP.CliXmlError;
+                string message = string.Format(
+                    null,
+                    template,
+                    isOutput ? MinishellStream.Output : MinishellStream.Error,
+                    Path,
+                    originalException.Message);
+                XmlException newException = new XmlException(
+                    message,
+                    originalException);
+
+                ErrorRecord error = new ErrorRecord(
+                    newException,
+                    "ProcessStreamReader_CliXmlError",
+                    ErrorCategory.SyntaxError,
+                    Path);
+                result.Add(new ProcessOutputObject(error, MinishellStream.Error));
+            }
 
-                            Host.BufferCell[,] bufferContents = this.Command.Context.EngineHostInterface.UI.RawUI.GetBufferContents(
-                                new Host.Rectangle(startPosition, endPosition));
+            return result;
+        }
 
-                            StringBuilder lineContents = new StringBuilder();
-                            StringBuilder bufferText = new StringBuilder();
+        private void InitOutputQueue()
+        {
+            //if output is redirected, start reading output of process in queue.
+            if (_startInfo.RedirectStandardOutput || _startInfo.RedirectStandardError)
+            {
+                lock (_sync)
+                {
+                    if (!_stopped)
+                    {
+                        _nativeProcessOutputQueue = new ConcurrentQueue<ProcessOutputObject>();
 
-                            for (int row = 0; row < bufferContents.GetLength(0); row++)
+                        if (_startInfo.RedirectStandardOutput)
+                        {
+                            bool isFirstOutput = true;
+                            bool isXmlCliOutput = false;
+                            _nativeProcess.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
```

Comment:
Yes, I know #1908 is not covered by this PR. My point is that we would still need a thread worker to transfer byte streams between native commands, so probably we should keep the `ProcessOutputReader` implementation there for now just in case that we need to reuse some of that code.
And for the same reason, using `OutputDataReceived` and `ErrorDataReceived` to collect process output/error may not be appropriate since we probably have to change it again using a thread worker when addressing #1908